src/huggingface_hub/constants.py

```diff
@@ -126,12 +128,40 @@ def _as_int(value: Optional[str]) -> Optional[int]:
 default_assets_cache_path = os.path.join(HF_HOME, "assets")
 
 # Legacy env variables
-HUGGINGFACE_HUB_CACHE = os.getenv("HUGGINGFACE_HUB_CACHE", default_cache_path)
-HUGGINGFACE_ASSETS_CACHE = os.getenv("HUGGINGFACE_ASSETS_CACHE", default_assets_cache_path)
+HUGGINGFACE_HUB_CACHE = os.path.expandvars(
+    os.path.expanduser(
+        os.getenv(
+            "HUGGINGFACE_HUB_CACHE",
+            default_cache_path,
+        )
+    )
+)
+HUGGINGFACE_ASSETS_CACHE = os.path.expandvars(
+    os.path.expanduser(
+        os.getenv(
+            "HUGGINGFACE_ASSETS_CACHE",
+            default_assets_cache_path,
+        )
+    )
+)
```

Comment:
```suggestion
HUGGINGFACE_HUB_CACHE = os.getenv("HUGGINGFACE_HUB_CACHE", default_cache_path)
HUGGINGFACE_ASSETS_CACHE = os.getenv("HUGGINGFACE_ASSETS_CACHE", default_assets_cache_path)
```

I wouldn't change the logic for legacy variables as we don't want to improve support for them (if a user really needs to expand variables, they can just use the correct env variables `HF_HUB_CACHE`/`HF_ASSETS_CACHE`